src/cpp/src/sampler.cpp

```diff
@@ -770,152 +907,61 @@ SamplerOutput Sampler::sample(const std::vector<SequenceGroup::Ptr> & sequence_g
-                                continue;
-                            }
-                            // update log prob just while validation process
-                            if (!is_extend_sequence) {
-                                OPENVINO_ASSERT(generated_and_verified_len < running_sequences[running_sequence_id]->get_generated_len());
-                                running_sequence->update_generated_log_prob(generated_and_verified_len, sampled_token.m_log_prob);
-                            }
-                        }
-                        register_new_token(sampled_token, running_sequences[running_sequence_id], logit_processor, is_extend_sequence, is_validation_mode_enabled);
-                        // to exit from sampling in case of failed token validation
-                        if (!is_validation_passed) {
-                            break;
-                        }
-                    }
-                    min_generated_len = std::min(min_generated_len, running_sequence->get_generated_len());
-                }
-                align_all_sequence_len(sequence_group, min_generated_len, logit_processor);
-                for (const auto& dropped_seq_id : _try_finish_generation(sequence_group)) {
-                    sampler_output.m_dropped_sequences.push_back(dropped_seq_id);
-                }
-            } else if (sampling_params.is_beam_search()) {
-                uint64_t request_id = sequence_group->get_request_id();
-
-                // create beam search info if we are on the first generate
-                if (m_beam_search_info.find(request_id) == m_beam_search_info.end()) {
-                    m_beam_search_info.emplace(request_id, GroupBeamSearcher(sequence_group, m_tokenizer));
-                }
-
-                // current algorithm already adds new tokens to running sequences and
-                m_beam_search_info.at(request_id).select_next_tokens(sequence_group_logits, sampler_output, stop_strings);
-
-                // check max length stop criteria
-                std::vector<Sequence::Ptr> running_sequences = sequence_group->get_running_sequences();
-                if (!sequence_group->has_finished() &&
-                    running_sequences[0]->get_generated_len() == sampling_params.max_new_tokens) {
-                    // stop sequence by max_new_tokens
-                    m_beam_search_info.at(request_id).finalize(sampler_output);
-                }
-            }
-            // Notify handle after sampling is done. 
-            // For non-streaming this is effective only when the generation is finished.
-            OPENVINO_ASSERT(num_tokens_to_process >= max_removed_tokens_per_request);
-            sequence_group->notify_handle();
+            // Call sample_from_sequence_group asynchronously
+            sg_sampling_future_map[sequence_group] = m_thread_pool.submit(&Sampler::sample_from_sequence_group, this, sequence_group, sequence_group_logits,
```

Comment:
We have `logits_processor` and `stop_strings` inside `Sampler` obj. Why so we need to use it as an additional args of function??

Reply:
I wanted to avoid parallel access to first layer sampler internals. Accessing global logit processor and stop string maps from sampler would require synchronization in `sample_from_sequence_group` method. I wanted this method to be rather unaware of the higher level structures as much as possible so there would be no need for synchronization on this level.

Reply:
Minor: Because, we parallelize  sampling per sequence_group, looks like can take logit processor & stop_strings from sampler by safe approach..